### PR TITLE
Allow setting CN and validity days on certificate generation

### PR DESCRIPTION
**Pull request by lucasmrod**
This allow setting a custom CN and validity days on `GET /v1/tokenpki/{name}`. Such API was always using default values CN=depserver and days=1.

See https://github.com/micromdm/nanodep/pull/5#discussion_r927016919.

### CODE REVIEW

**Comment by jessepeterson** [comment]
In general this is great! Have you noticed an actual problem with the validity date? I recall using an expired certificate and it working on the ABM portal but can't remember clearly: I think the certificate is used purely as a vessel for the public key (for encryption) rather than any of its other attributes. 🤷

The only change I'd make is not requiring the parameters in the tools/shell script. They could be in the shell script - just not required parameters ($1 etc). I want to keep those as simple as possible for end users. The API docs (thanks!) specify the variables so for integrations that could work.

Thanks again!

**Comment by jessepeterson** [comment]
Actually thinking about it again - if those params are omitted they'll just be blank as they're passed on. Perhaps then just the QuickStart docs can omit them :)

**Comment by lucasmrod** [comment]
> Have you noticed an actual problem with the validity date? I recall using an expired certificate and it working on the ABM portal but can't remember clearly: I think the certificate is used purely as a vessel for the public key (for encryption) rather than any of its other attributes. 🤷

Oh! I would expect any tooling to fail if a certificate is expired :) (And thus I thought it would be handy to allow people to configure more than one day to allow for more flexibility, e.g. the user generating the certificate might belong to different team than user uploading it to ABM.)

> The only change I'd make is not requiring the parameters in the tools/shell script. They could be in the shell script - just not required parameters ($1 etc). I want to keep those as simple as possible for end users. The API docs (thanks!) specify the variables so for integrations that could work.

Good idea. I've set default arguments in the tools and amended the `docs/quickstart.md`.

Let me know if that works.

File: cmd/deptokens/main.go [hunk 1]
```diff
@@ -13,22 +13,21 @@ import (
...
-		flCert     = flag.String("cert", "cert.pem", "path to certificate")
-		flKey      = flag.String("key", "cert.key", "path to key")
-		flPassword = flag.String("password", "", "password to encrypt/decrypt private key with")
-		flTokens   = flag.String("token", "", "path to tokens")
-		flForce    = flag.Bool("f", false, "force overwriting the keypair")
-		flVersion  = flag.Bool("version", false, "print version")
+		flCert         = flag.String("cert", "cert.pem", "path to certificate")
+		flKey          = flag.String("key", "cert.key", "path to key")
+		flCN           = flag.String("cn", defaultCN, "common name to use when creating the certificate")
+		flValidityDays = flag.Int64("validity-days", 0, "validity of the certificate in days")
```
Comment (jessepeterson) [inline cmd/deptokens/main.go:26]: default to 1, and some nits

```suggestion
		flDays = flag.Int64("days", 1, "validity of the certificate in days")
```

File: cmd/deptokens/main.go [hunk 2]
```diff
@@ -13,22 +13,21 @@ import (
 	"github.com/micromdm/nanodep/tokenpki"
 )
 
-const (
-	defaultCN   = "deptokens"
-	defaultDays = 1
-)
+const defaultCN = "deptokens"
```
Comment (jessepeterson) [inline cmd/deptokens/main.go:16]: just use default in flags?

File: cmd/deptokens/main.go [hunk 3]
```diff
@@ -13,22 +13,21 @@ import (
...
 	var (
-		flCert     = flag.String("cert", "cert.pem", "path to certificate")
-		flKey      = flag.String("key", "cert.key", "path to key")
-		flPassword = flag.String("password", "", "password to encrypt/decrypt private key with")
-		flTokens   = flag.String("token", "", "path to tokens")
-		flForce    = flag.Bool("f", false, "force overwriting the keypair")
-		flVersion  = flag.Bool("version", false, "print version")
+		flCert         = flag.String("cert", "cert.pem", "path to certificate")
+		flKey          = flag.String("key", "cert.key", "path to key")
+		flCN           = flag.String("cn", defaultCN, "common name to use when creating the certificate")
```
Comment (jessepeterson) [inline cmd/deptokens/main.go:25]:
```suggestion
		flCN           = flag.String("cn", "deptokens", "common name to use when creating the certificate")
```

File: docs/openapi.yaml [hunk 4]
```diff
@@ -158,6 +158,19 @@ paths:
       description: Generate and store a new X.509 certificate and RSA private key (keypair) for exchanging the encrypted DEP OAuth1 tokens via the Apple ABM/ASM/BE portal. Each request generates a new (and overwrites the existing) keypair. The certificate is returned.
       security:
         - basicAuth: []
+      parameters:
```
Comment (jessepeterson) [inline docs/openapi.yaml:161]: ❤️ thanks for updating this, too!

File: docs/operations-guide.md [hunk 5]
```diff
@@ -172,7 +172,7 @@ For the DEP "MDM server" in the environment variable $DEP_NAME (see above) this
 ##### Example usage
```
Comment (jessepeterson) [inline docs/operations-guide.md:172]: Let's add a note like some of the other scripts that describes what the options do.

Let's also add a (very brief) note in the API endpoint in the ops guide, too.

File: http/api/tokenpki.go [hunk 6]
```diff
@@ -55,7 +55,24 @@ func GetCertTokenPKIHandler(store TokenPKIStorer, logger log.Logger) http.Handle
 			return
 		}
 		logger = logger.With("name", r.URL.Path)
-		key, cert, err := tokenpki.SelfSignedRSAKeypair(defaultCN, defaultDays)
+		daysArg := r.URL.Query().Get("validity_days")
+		if daysArg == "" {
+			logger.Info("msg", "validity_days check", "err", "missing validity_days")
+			http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
+			return
```
Comment (jessepeterson) [inline http/api/tokenpki.go:62]: Needs to be optional with a sane default.

File: docs/openapi.yaml [hunk 7]
```diff
@@ -158,6 +158,19 @@ paths:
...
+      - in: query
+        name: cn
+        required: false
+        schema:
+          type: string
+          example: "depserver"
+      - in: query
+        name: validity_days
+        required: true
```
Comment (jessepeterson) [inline docs/openapi.yaml:170]: Should be optional.

File: tokenpki/cert.go [hunk 8]
```diff
@@ -14,7 +14,7 @@ import (
 // SelfSignedRSAKeypair generates a 2048-bit RSA private key and self-signs an
 // X.509 certificate using it. You can set the Common Name in cn and the
 // validity duration with days.
-func SelfSignedRSAKeypair(cn string, days int) (*rsa.PrivateKey, *x509.Certificate, error) {
+func SelfSignedRSAKeypair(cn string, days int64) (*rsa.PrivateKey, *x509.Certificate, error) {
```
Comment (jessepeterson) [inline tokenpki/cert.go:17]: That's.. a lot of days. :) Curious why the change?

File: tools/cfg-get-cert.sh [hunk 9]
```diff
@@ -1,6 +1,8 @@
 #!/bin/sh
 
-URL="${BASE_URL}/v1/tokenpki/${DEP_NAME}"
+CN="${1:-depserver}"
+VALIDITY_DAYS="${2:-1}"
+URL="${BASE_URL}/v1/tokenpki/${DEP_NAME}?cn=$CN&validity_days=$VALIDITY_DAYS"
```
Comment (jessepeterson) [inline tools/cfg-get-cert.sh:5]: Any reason not to keep it as you had it? If both args are optional they'll just be submitted as empty which should use defaults on the server side. :)

```suggestion
URL="${BASE_URL}/v1/tokenpki/${DEP_NAME}?cn=$1&validity_days=$2"
```

**Review by jessepeterson — CHANGES_REQUESTED** [review]
In addition to the other comments had a few code questions/comments. Thanks for this work!

File: tokenpki/cert.go (hunk 8 above)
Comment (lucasmrod) [reply tokenpki/cert.go:17]: I've changed the `days` type to `int64` because it's being converted to a `time.Duration` here:
https://github.com/micromdm/nanodep/blob/077b73dfe75f5cf5ce12601ce1ca59bbb925bb0e/tokenpki/cert.go#L29

And `time.Duration`'s underlying type is `int64`.

**Review by jessepeterson — APPROVED** [review]
lgtm! thanks for the changes (and sorry for the delay!)
